File: rxauth/src/main/java/com/sebastianmarschall/rxsmartlock/observable/RetrieveCredentialObservable.java
package com.sebastianmarschall.rxsmartlock.observable;


import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.credentials.Credential;
import com.google.android.gms.auth.api.credentials.CredentialRequest;
import com.google.android.gms.auth.api.credentials.CredentialRequestResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.sebastianmarschall.rxsmartlock.exception.ConnectionException;
import com.sebastianmarschall.rxsmartlock.exception.ConnectionSuspendedException;
import com.sebastianmarschall.rxsmartlock.exception.StatusException;

import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.functions.Cancellable;

public class RetrieveCredentialObservable implements SingleOnSubscribe<Credential> {

    private Context mContext;
    private CredentialRequest mCredentialRequest;

    public RetrieveCredentialObservable(Context context, CredentialRequest request) {
        mContext = context;
        mCredentialRequest = request;
    }

    @Override
    public void subscribe(SingleEmitter<Credential> subscriber) throws Exception {

        final GoogleApiClient googleApiClient = buildGoogleApiClient(subscriber);

        try {
            googleApiClient.connect();
        } catch (Exception e) {
            subscriber.onError(e);
        }

        subscriber.setCancellable(new Cancellable() {
            @Override
            public void cancel() throws Exception {
                if (googleApiClient.isConnected() || googleApiClient.isConnecting()) {
                    googleApiClient.disconnect();
                }
            }
        });
    }

    private GoogleApiClient buildGoogleApiClient(SingleEmitter<? super Credential> observer) {

        GoogleApiClientCallbacks clientCallbacks = new GoogleApiClientCallbacks(observer);
        GoogleApiClient client = new GoogleApiClient.Builder(mContext)
                .addApi(Auth.CREDENTIALS_API)
                .addConnectionCallbacks(clientCallbacks)
                .addOnConnectionFailedListener(clientCallbacks)
                .build();
        clientCallbacks.setGoogleApiClient(client);
        return client;

    }

    private class GoogleApiClientCallbacks implements GoogleApiClient.ConnectionCallbacks,
            GoogleApiClient.OnConnectionFailedListener, ResultCallback<CredentialRequestResult> {

        private GoogleApiClient googleApiClient;
        private SingleEmitter<? super Credential> subscriber;

        public GoogleApiClientCallbacks(SingleEmitter<? super Credential> observer) {
            subscriber = observer;
        }

        @Override
        public void onConnected(Bundle bundle) {

            try {
                Auth.CredentialsApi.request(googleApiClient, mCredentialRequest)
                        .setResultCallback(this);
            } catch (Exception e) {
                subscriber.onError(e);
            }

        }

        @Override
        public void onConnectionSuspended(int i) {
            subscriber.onError(new ConnectionSuspendedException(i));
        }

        @Override
        public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
            subscriber.onError(new ConnectionException(connectionResult));
        }

        @Override
        public void onResult(@NonNull CredentialRequestResult credentialRequestResult) {

            Status status = credentialRequestResult.getStatus();
            if (status.isSuccess()) {
                onCredentialRetrieved(credentialRequestResult.getCredential());
            } else {
                resolveResult(status);
            }

        }

        private void onCredentialRetrieved(Credential credential) {
            subscriber.onSuccess(credential);
        }

        private void resolveResult(Status status) {
            subscriber.onError(new StatusException(status));
        }

        public void setGoogleApiClient(GoogleApiClient googleApiClient) {
            this.googleApiClient = googleApiClient;
        }

    }

}

File: rxauth/src/main/java/com/sebastianmarschall/rxsmartlock/observable/RetrieveCredentialObservable.kt
package com.sebastianmarschall.rxsmartlock.observable


import android.content.Context
import android.os.Bundle
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.credentials.Credential
import com.google.android.gms.auth.api.credentials.CredentialRequest
import com.google.android.gms.auth.api.credentials.CredentialRequestResult
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.common.api.Status
import com.sebastianmarschall.rxsmartlock.exception.ConnectionException
import com.sebastianmarschall.rxsmartlock.exception.ConnectionSuspendedException
import com.sebastianmarschall.rxsmartlock.exception.StatusException
import io.reactivex.SingleEmitter
import io.reactivex.SingleOnSubscribe

class RetrieveCredentialObservable(private val mContext: Context, private val mCredentialRequest: CredentialRequest) : SingleOnSubscribe<Credential> {

    override fun subscribe(subscriber: SingleEmitter<Credential>) {

        val googleApiClient = buildGoogleApiClient(subscriber)

        try {
            googleApiClient.connect()
        } catch (e: Exception) {
            subscriber.onError(e)
        }

        subscriber.setCancellable {
            if (googleApiClient.isConnected || googleApiClient.isConnecting) {
                googleApiClient.disconnect()
            }
        }
    }

    private fun buildGoogleApiClient(observer: SingleEmitter<in Credential>): GoogleApiClient {

        val clientCallbacks = GoogleApiClientCallbacks(observer)
        val client = GoogleApiClient.Builder(mContext)
                .addApi(Auth.CREDENTIALS_API)
                .addConnectionCallbacks(clientCallbacks)
                .addOnConnectionFailedListener(clientCallbacks)
                .build()
        clientCallbacks.setGoogleApiClient(client)
        return client

    }

    private inner class GoogleApiClientCallbacks(private val subscriber: SingleEmitter<in Credential>) : GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener, ResultCallback<CredentialRequestResult> {

        private var googleApiClient: GoogleApiClient? = null

        override fun onConnected(bundle: Bundle?) {

            try {
                Auth.CredentialsApi.request(googleApiClient, mCredentialRequest)
                        .setResultCallback(this)
            } catch (e: Exception) {
                subscriber.onError(e)
            }

        }

        override fun onConnectionSuspended(i: Int) {
            subscriber.onError(ConnectionSuspendedException(i))
        }

        override fun onConnectionFailed(connectionResult: ConnectionResult) {
            subscriber.onError(ConnectionException(connectionResult))
        }

        override fun onResult(credentialRequestResult: CredentialRequestResult) {

            val status = credentialRequestResult.status
            if (status.isSuccess) {
                onCredentialRetrieved(credentialRequestResult.credential)
            } else {
                resolveResult(status)
            }

        }

        private fun onCredentialRetrieved(credential: Credential) {
            subscriber.onSuccess(credential)
        }

        private fun resolveResult(status: Status) {
            subscriber.onError(StatusException(status))
        }

        fun setGoogleApiClient(googleApiClient: GoogleApiClient) {
            this.googleApiClient = googleApiClient
        }

    }

}
